lib/solver.ml

```diff
@@ -204,21 +137,58 @@ and solve_sub sol ty1 ty2 =
   | TyFun (args1, _), TyFun (args2, _) ->
      let filename = "TODO:filename" in
      let line = -1 (*TODO:line*) in
-     let actual = ty1 in
-     let expected = ty2 in
+     let actual = ty1' in
+     let expected = ty2' in
      let message = "the fun args are not different length" in
      Error Known_error.(FialyzerError(TypeError {filename; line; actual; expected; message}))
   | TyUnion (ty11, ty12), _ ->
      let open Result in
      solve_sub sol ty11 ty2 >>= fun sol' -> solve_sub sol' ty12 ty2
   | _ ->
      (* TODO: normalize *)
-     if is_subtype sol ty1 ty2 then
+     if is_subtype ty1' ty2' then
        Ok sol
      else
        let filename = "TODO:filename" in
        let line = -1 (*TODO:line*) in
-       let actual = ty1 in
-       let expected = ty2 in
-       let message = "there is no solution that satisfies subtype constraints" in
+       let actual = ty1' in
+       let expected = ty2' in
+       let message = !%"there is no solution that satisfies subtype constraints" in
        Error Known_error.(FialyzerError(TypeError {filename; line; actual; expected; message}))
+
+let solve_eq sol ty1 ty2 =
+  let open Result in
+  solve_sub sol ty1 ty2 >>= fun sol' ->
+  solve_sub sol' ty2 ty1
+
+let rec solve1 sol = function
+  | Empty -> Ok sol
+  | Eq (ty1, ty2) ->
+     solve_eq sol ty1 ty2
+  | Subtype (ty1, ty2) ->
+     solve_sub sol ty1 ty2
+  | Conj cs ->
+     solve_conj sol cs
+  | Disj cs ->
+     Error Known_error.(FialyzerError (NotImplemented {issue_link="https://github.com/dwango/fialyzer/issues/99"}))
+and solve_conj sol = function
+  | [] -> Ok sol
+  | c :: cs ->
+     let open Result in
+     solve1 sol c >>= fun sol' ->
+     solve_conj sol' cs
+
+let rec solve sol cs =
+  let open Result in
+  solve1 sol cs >>= fun sol' ->
+  let updated_vars =
```

Comment:
Can we use `Map.equal` or `Map.symmetric_diff`?

Reply:
Good idea! Fixed!

Reply:
👍